packages/report-server/src/reportBuilder/transform/aliases/summaryAliases.ts

```diff
@@ -0,0 +1,77 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2020 Beyond Essential Systems Pty Ltd
+ */
+
+import { Row } from '../../types';
+
+/**
+ * Inserts a new column with a value that is the calculation of the percentage of 1s out of 1s and 0s in a row.
+ * { rowFieldName: 'Availability of Male Condoms', facilityNameA: 'N', facilityNameB: 'N', facilityNameC: 'Y', facilityNameD: 'N'  }
+ *  => { rowFieldName: 'Availability of Male Condoms', facilityNameA: 'N', facilityNameB: 'N', facilityNameC: 'Y', facilityNameD: 'N', rowSummary: 0.75 }
+ */
+
+const getRowsWithSummaryColumn = (rows: Row[]) => {
+  return rows.map(row => {
+    const numerator = Object.entries(row).filter(([key,value]) => value === 'N').length;
+    const denominator = Object.entries(row).filter(([key,value]) => value === 'N' || value === 'Y').length;
+    if (denominator === 0) {
+      return { 'summaryColumn': undefined, ...row };
+    }
+    const summary = `${(numerator / denominator * 100).toFixed(1)}%`;
+    return { 'summaryColumn': summary, ...row };
+  })
+};
+
+const getFractionRow = (rows: Row[],isNumerator: boolean) => {
+  return rows.reduce((accum: Row, currentRow: Row) => {
+    const rowToArray = Object.entries(currentRow)
+    let futureValue: Row = {...accum}
+    rowToArray.forEach(([key,value]) => {
+      if (isNumerator) {
+        if (value === 'N') {
+          if (typeof futureValue[key] === 'number') {
+            futureValue[key] = futureValue[key] + 1;
+          } else {
+            futureValue[key] = 1;
+          }
+        } else if (value === 'Y'){
+          if (typeof futureValue[key] === 'number') {
+            futureValue[key] = futureValue[key] + 0;
+          } else {
+            futureValue[key] = 0;
+          }
+        }
+      } else {
+        if (value === 'N' || value === 'Y') {
+          if (typeof futureValue[key] === 'number') {
+            futureValue[key] = futureValue[key] + 1;
+          } else {
+            futureValue[key] = 1;
+          }
+        }
+      }
+    })
+    return futureValue
+  },{})
+};
+
+const getSummaryRow = (numeratorRow: Row, denominatorRow: Row) => {
+  return Object.fromEntries(Object.entries(denominatorRow).map(([key,value]) => {
+    const numeratorValue = numeratorRow[key] as number;
+    if (typeof value === 'number') {
+      const percentage = `${(numeratorValue / value * 100).toFixed(1)}%`;
+    return [key, percentage]
+    }
+    return [key, value]
+  }))
+}
+
+export const insertSummaryRowAndColumn = () => (rows: Row[]) => {
+  const rowsWithSummaryColumn = getRowsWithSummaryColumn(rows);
+  const numeratorRow = getFractionRow(rows,true);
+  const denominatorRow = getFractionRow(rows,false)
+  const summaryRow = getSummaryRow(numeratorRow,denominatorRow)
+  return [...rowsWithSummaryColumn,summaryRow]
```

Comment:
Apologies as I'm being nit picky here, and feel free to push back on this if you'd rather just push this through, however I felt you might appreciate my perspective here?

Tbh I find the logical flow of the algorithm here a bit unexpected. We're presented with a set of rows, which we plan on adding a summary column to each, and then finally adding a summary row to the table. Knowing which columns we are summarising seems to be a key concept here, but it's sort of handled implicitly in the if/else logic of `getRowsWithSummaryColumn()` and `getFractionRow()`.

Knowing which columns to summarise does feel a touch tricky, but a safeish bet seems to be all columns whose values are all either `'Y' | 'N' | undefined`? That's not technically what we have here, as we do it on a cell by cell basis which should largely be fine, however it could lead to small bugs in some circumstances (I recently did a matrix where we had a column whose value was the first letter of the medicine, which could be `'Y'` or `'N'`).

I think the way I intuit this algorithm would be based on how I might do this process manually, and goes a bit like:
```
const columnsToSummarise = detectColumnsToSummarise(rows); // Works out which columns are all Y/N/undefined
const rowsWithSummaryColumn = rows.map(row => addSummaryColumn(row, columnsToSummarise));
const summaryRow = getSummaryRow(rows, columnsToSummarise); 
return [...rowsWithSummaryColumn, summaryRow];
```

Very happy to chat on this as design ideas like this can be very subjective 👍